src/Analyzers/Core/Analyzers/NamespaceSync/AbstractNamespaceSyncDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Analyzers.NamespaceSync
+{
+    internal abstract class AbstractNamespaceSyncDiagnosticAnalyzer<TNamespaceSyntax> : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+        where TNamespaceSyntax : SyntaxNode
+    {
+        public const string RootNamespaceOption = "build_property.RootNamespace";
+        public const string ProjectDirOption = "build_property.ProjectDir";
+
+        public AbstractNamespaceSyncDiagnosticAnalyzer(LocalizableResourceString title, LocalizableResourceString message)
+            : base(IDEDiagnosticIds.NamespaceSyncAnalyzerDiagnosticId,
+                EnforceOnBuild.Recommended,
+                (IPerLanguageOption?)null,
```

Comment:
Pass in `CSharpCodeStyleOptions.PreferNamespaceMatchFolderStructure` from derived type ctor.